File: plsql-checks/src/main/java/br/com/felipezorzo/sonar/plsql/checks/IfWithExitCheck.java
package br.com.felipezorzo.sonar.plsql.checks;

import org.sonar.api.server.rule.RulesDefinition;
import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.squidbridge.annotations.ActivatedByDefault;
import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;

import com.sonar.sslr.api.AstNode;

import br.com.felipezorzo.sonar.plsql.api.PlSqlGrammar;
import br.com.felipezorzo.sonar.plsql.api.PlSqlKeyword;

@Rule(
    key = IfWithExitCheck.CHECK_KEY,
    priority = Priority.MINOR
)
@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.READABILITY)
@SqaleConstantRemediation("5min")
@ActivatedByDefault
public class IfWithExitCheck extends BaseCheck {
    
    public static final String CHECK_KEY = "IfWithExit";

    @Override
    public void init() {
        subscribeTo(PlSqlGrammar.EXIT_STATEMENT);
    }

    @Override
    public void visitNode(AstNode node) {
        AstNode statement = node.getParent();
        AstNode ifStatement = statement.getParent();
        if (ifStatement.is(PlSqlGrammar.IF_STATEMENT) &&
            !ifStatement.hasDirectChildren(PlSqlKeyword.ELSIF, PlSqlKeyword.ELSE) &&
            ifStatement.getChildren(PlSqlGrammar.STATEMENT).size() == 1) {
            getContext().createLineViolation(this, getLocalizedMessage(CHECK_KEY), node);
        }
    }

}

File: zpa-checks/src/main/kotlin/com/felipebz/zpa/checks/IfWithExitCheck.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2026 Felipe Zorzo
 * mailto:felipe AT felipezorzo DOT com DOT br
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.felipebz.zpa.checks

import com.felipebz.zpa.api.PlSqlGrammar
import com.felipebz.zpa.api.annotations.*
import com.felipebz.zpa.api.syntax.IfStatement
import com.felipebz.zpa.api.syntax.SyntaxViews

@Rule(priority = Priority.MINOR)
@ConstantRemediation("5min")
@RuleInfo(scope = RuleInfo.Scope.ALL)
@ActivatedByDefault
@OptIn(ZpaExperimentalApi::class)
class IfWithExitCheck : AbstractBaseCheck() {

    override fun init() {
        subscribeTo(SyntaxViews.IF_STATEMENT, ::visitIfStatement)
    }

    private fun visitIfStatement(statement: IfStatement) {
        if (statement.elsifBranches.isEmpty() && statement.elseBranch == null &&
                statement.statementAstNodes.singleOrNull()?.hasDirectChildren(PlSqlGrammar.EXIT_STATEMENT) == true) {
            addIssue(statement, getLocalizedMessage())
        }
    }

}
